libdevcore/db.h

```diff
@@ -21,16 +21,80 @@
 
 #pragma once
 
+#include "Exceptions.h"
 #include "dbfwd.h"
 
-#pragma warning(push)
-#pragma warning(disable: 4100 4267)
-#if ETH_ROCKSDB
-#include <rocksdb/db.h>
-#include <rocksdb/write_batch.h>
-#else
-#include <leveldb/db.h>
-#include <leveldb/write_batch.h>
-#endif
-#pragma warning(pop)
-#define DEV_LDB 1
+#include <memory>
+#include <string>
+
+namespace dev
+{
+namespace db
+{
+
+// Transaction implements database transaction for a specific concrete database
+// implementation. The transaction implementation must call rollback in its
+// destructor if commit has not yet been called (making sure to catch any
+// exceptions).
+class Transaction
+{
+public:
+	virtual ~Transaction() = default;
+
+	virtual void insert(Slice const& _key, Slice const& _value) = 0;
+	virtual void kill(Slice const& _key) = 0;
+
+	virtual void commit() = 0;
+	virtual void rollback() = 0;
+
+protected:
+	Transaction() = default;
+	// Noncopyable
+	Transaction(const Transaction&) = delete;
+	Transaction& operator=(const Transaction&) = delete;
+	// Movable
+	Transaction(Transaction&&) = default;
```

Comment:
Does it need to be movable? I don't see move constructor defined in `LevelDBTransaction`, I think I'd leave it nonmovable

Reply:
Nonmovable doesn't really make sense to me. If anything, I would say the default should be movable until proven otherwise.

Reply:
Some evidence to back this up: https://stackoverflow.com/a/14304618/1930331.